hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -3684,32 +3625,96 @@ class Emit[C](val ctx: EmitContext, val cb: EmitClassBuilder[C]) {
       cb.memoize(cb.invokeCode[Boolean](sort, cb.this_, region, l, r))
   }
 
-  def emitLet[A](
-    emitI: (IR, EmitCodeBuilder, EmitEnv) => IEmitCode,
-    emitBody: (IR, EmitCodeBuilder, EmitEnv) => A,
+  def emitLetBindings(
+    emitI: (IR, EmitCodeBuilder, EmitEnv, Value[Region]) => IEmitCode,
+    emitVoid: (IR, EmitCodeBuilder, EmitEnv, Value[Region]) => Unit,
   )(
     let: Let,
     cb: EmitCodeBuilder,
     env: EmitEnv,
-  ): A = {
+    r: Value[Region],
+  ): EmitEnv = {
     val uses: mutable.Set[String] =
       ctx.usesAndDefs.uses.get(let) match {
         case Some(refs) => refs.map(_.t.name)
         case None => mutable.Set.empty
       }
 
-    emitBody(
-      let.body,
-      cb,
-      let.bindings.foldLeft(env) { case (newEnv, (name, ir)) =>
-        if (!uses.contains(name)) newEnv
-        else {
-          val value = emitI(ir, cb, newEnv)
+    def emitChunk(cb: EmitCodeBuilder, bindings: Seq[(String, IR)], env: EmitEnv, r: Value[Region])
+      : EmitEnv =
+      bindings.foldLeft(env) { case (newEnv, (name, ir)) =>
+        if (ir.typ == TVoid) {
+          emitVoid(ir, cb, newEnv, r)
+          newEnv
+        } else if (IsPure(ir) && !uses.contains(name)) {
+          newEnv
+        } else {
+          val value = emitI(ir, cb, newEnv, r)
           val memo = cb.memoizeMaybeStreamValue(value, s"let_$name")
           newEnv.bind(name, memo)
         }
-      },
-    )
+      }
+
+    @tailrec def go(
+      env: EmitEnv,
+      chunkStart: Int,
+      pos: Int,
+      chunkSize: Int,
+      groupIdx: Int,
+    ): EmitEnv = {
+
+      def emitChunkWrapped(): EmitEnv = {
+        val mb = cb.emb.genEmitMethod(
+          s"begin_group_$groupIdx",
+          FastSeq[ParamType](classInfo[Region]),
+          UnitInfo,
+        )
+        var newEnv = env
```

Comment:
This env needs to be restricted to fields and formal parameters only.
We may not bind any locals at present, but I can imagine if we did and they were used across a split, we'd get a nasty runtime error

Reply:
Also, make this a param of `emitChunkWrapped` - this ref copy just looks ugly

Reply:
> This env needs to be restricted to fields and formal parameters only. We may not bind any locals at present, but I can imagine if we did and they were used across a split, we'd get a nasty runtime error

The env must have this property already, as we currently emit parts of the ir into separate methods, and simply pass in the current env. See for instance `Emit.emitInSeparateMethod`.

> Also, make this a param of `emitChunkWrapped` - this ref copy just looks ugly

I'm not sure what you mean. The `newEnv` var is so that we can smuggle out the update env from `emitChunk` out of `voidWithBuilder`.

Reply:
Oh I see.